### ISSUE TITLE
[BUG] NVTabular data loader for TensorFlow validation is slow

### ISSUE THREAD

**Issue by bschifferer** [NVIDIA/NVTabular#462]
**Describe the bug**
Using NVTabular data loader for TensorFlow for validation with criteo dataset is slow:

```
    validation_callback = KerasSequenceValidater(valid_dataset_tf)
    history = model.fit(train_dataset_tf, 
                        epochs=EPOCHS, 
                        steps_per_epoch=20, callbacks=[validation_callback])
```
Training : 2min for 2288 steps 
Validation: Estimated 55min for 3003 steps
Same batch-size, dataset, etc. ... **The validation dataset is 1.3x bigger, but iterating through the validation dataset takes 27x more time than for training.**

**Steps/Code to reproduce bug**
An example is provided here: https://github.com/bschifferer/NVTabular/blob/criteo_tf_slow/examples/criteo_tensorflow_slow.ipynb

You notice that iterating over the training dataset takes in average ~1sec per 10 batches but it takes 4-6s per 10 batches in the validation dataset.

**Expected behavior**
Validation loop should be similar fast than training loop

**Additional context**
There are multiple hypothesis and tests:
* This behavior can be observed by just iterating over the dataset and execute the forward-pass of the model
* If we switch training/validation dataloader, then the validation dataloader is fast and the training dataloader is slow. Meaning, that a iteration over the 2nd data loader. **Hypothesis is that the GPU memory is not released from the 1st data loader and blocks the pipeline**
* If we remove the forward-pass of the model in the loop, then both iterations are fast. It has probably something to do with moving the data to the TensorFlow model
* I tried out using `tf.keras.backend.clear_session()` between the iterations, but did not help
* I tried out between the iterations
```
from numba import cuda
cuda.select_device(0)
cuda.close()
```
but resulted in an error
* I tried out to use separate subprocesses, but did not improve performance

**Comment by rnyak** [comment]
@bschifferer I guess your hypothesis `the GPU memory is not released from the 1st data loader and blocks the pipeline` makes sense.  AFAIK, there is no step/script to release GPU memory.

**Comment by EvenOldridge** [comment]
@jperez999 we think this is happening in the TF dataloader only.  @bschifferer will confirm by testing PyTorch.  Can you take a look please.

**Comment by jperez999** [comment]
@bschifferer I was unable to reproduce your results. I ran this on a V100 32gb GPU and I got the following output for each of the dataloader analysis runs...
![Screen Shot 2020-12-02 at 4 10 48 PM](https://user-images.githubusercontent.com/37191411/100932152-4f03e880-34a0-11eb-8364-567cd7fc3977.png)
![Screen Shot 2020-12-02 at 4 10 56 PM](https://user-images.githubusercontent.com/37191411/100932156-50cdac00-34a0-11eb-8457-34c206111c27.png)
![Screen Shot 2020-12-02 at 4 11 15 PM](https://user-images.githubusercontent.com/37191411/100932165-53c89c80-34a0-11eb-8daa-9888051efdb3.png)


None of my runs experience the slowdown that should be captured in these runs. 

The only change I made to your notebook was to update the input path and activated preprocessing.

**Comment by bschifferer** [comment]
@jperez999 - thanks for testing. I still have the bug.

I used the latest nvtabular:0.3 container
```docker run --gpus 1 --cap-add SYS_PTRACE -it -p 8000:8000 -p 8888:8888 -v /raid/:/raid/ -v /workspace:/workspace nvcr.io/nvidia/nvtabular:0.3 /bin/bash```
GPU:
![image](https://user-images.githubusercontent.com/19567274/101152230-0c1e4e00-3623-11eb-85be-5a81c4faa6b8.png)

Versions:
nvtabular: 0.2.0+66.gbc91a03
cudf: 0+untagged.1.ga6296e3
tensorflow: 2.5.0-dev20201121
rmm: has no version call
python 3.7.8

BUG:
![image](https://user-images.githubusercontent.com/19567274/101152273-1f311e00-3623-11eb-9b8d-e3146859e921.png)

**Comment by bschifferer** [comment]
@jperez999 if I run it on NGC, I do NOT have the same behavior/bug.

![image](https://user-images.githubusercontent.com/19567274/101165287-adfb6600-3636-11eb-9f39-abf33190245b.png)

It is the same container. I do not have raid system their either. Is this related to Tesla vs. Quadro?

**Comment by rnyak** [comment]
@bschifferer @jperez999 I am seeing similar behavior on NGC and on DGX06 with single V100-32 GB for validation with my code using full criteo dataset. Validation is very slow, and the strange part is that validation auc metric is not even print out  after training is done, even though I set it in the callbacks as:

```
validation_callback = KerasSequenceValidater(dataset)
model.fit(train_dataset, epochs=1, steps_per_epoch=100, callbacks=callbacks)
```
I tested the same code with only one day criteo set (day_0.parquet) and I can get the validation metric printed out after training. But validation is still slow.

**Comment by jperez999** [comment]
@rnyak, can we get a full report on this behavior your seeing (i.e. get us environment information) along with testing output. I think this may have to do with spilling and the size restraints on your instance on ngc. Also can you verify that you were running the 0.16 cudf release... I think 0.2 was locked to cudf 0.15
@bschifferer ok if you are not hitting this on NGC, I think this may have to do with the amount of host memory and the spilling occurring during dataloading... Going to run some testing to see if I can see this behavior on my end and will report back.

**Comment by rnyak** [comment]
@jperez999 thanks for looking into that. I switched to `nvidia/nvtabular:0.3` image on NGC. So the conda environment in NVT image is: `/opt/conda/envs/rapids`. This is a 2x V100-32 GB instance on NGC.

`cudf version: 0+untagged.1.ga6296e3`

my ngc job submission script is:

```
ngc batch run --name "tfdlrm_nvt" --preempt RUNONCE --ace nv-us-west-2 --instance dgx1v.32g.2.norm --commandline "pip install ipykernel && python -m ipykernel install --user --name rapids --display-name rapids && pip install jupyterlab; jupyter lab --ip=0.0.0.0 --allow-root --no-browser --NotebookApp.token='' --notebook-dir=/ --NotebookApp.allow_origin='*' & date; sleep 10d" --result /result --image "nvidia/nvtabular:0.3" --org nvidian --team prj-recsys --datasetid xxx:/dataset/ --workspace xxxxx:/workspace:RW --port 6006 --port 8787 --port 8888
```
I am using [accelerating-tensorflow.ipynb](https://github.com/NVIDIA/NVTabular/blob/main/examples/tensorflow/accelerating-tensorflow.ipynb) as example notebook with full criteo dataset.

If I run the notebook with validation set, my output is as below (please note that - val_loss: xxx - val_auroc: xxx  results are not printed out but it should be). Also  wall time is ~17 min.
```
Embedding parameter count: 101176680
Non-embedding parameter count: 2747145
[<callbacks.ThroughputLogger object at 0x7f13d0149f10>, <tensorflow.python.keras.callbacks.TensorBoard object at 0x7f13d01f8650>, <nvtabular.loader.tensorflow.KerasSequenceValidater object at 0x7f13d0046ad0>]
200/200 [==============================] - 196s 682ms/step - loss: 0.1610 - auroc: 0.6639
CPU times: user 8min 40s, sys: 11min 39s, total: 20min 19s
Wall time: 17min 13s

```
if I run without validation set, training completes in  1 min 24 sec. Huge difference:

```
Embedding parameter count: 101176680
Non-embedding parameter count: 2747145
[<callbacks.ThroughputLogger object at 0x7f70c04d4590>, <tensorflow.python.keras.callbacks.TensorBoard object at 0x7f70c062cb50>]
200/200 [==============================] - 61s 223ms/step - loss: 0.1601 - auroc: 0.6620
CPU times: user 43.4 s, sys: 3min 2s, total: 3min 45s
Wall time: 1min 24s
```

**Comment by bschifferer** [comment]
@rnyak I used the same ngc container and the same instance type. 

1. If it is about spilling from host memory to disk, maybe you can try out a larger machine to have more host memory
2. Is the data to a local file system? If the data is read from the a "dataset", then this can be slow
3. I use the same batch-size and most of the same functions, but I use a different network architecture as accelerating-tensorflow.ipynb . I think my network architecture is smaller, maybe thats the reason we see different behavior?

**Comment by rnyak** [comment]
@bschifferer  I am seeing this behavior both on NGC instance, and also on DGX06 system with single V100-32 GB GPU. My issue is not only about timing, it is also why I cannot have validation auc metric printed out, if I am using `validation_callback = KerasSequenceValidater(dataset)`. This is a strange behavior was not happening previously.

1. If it is about spilling from host memory to disk, maybe you can try out a larger machine to have more host memory.

**I can try that. But not sure about spilling.**

2. Is the data to a local file system? If the data is read from the a "dataset", then this can be slow.

**No, on NGC, it is not read from the `dataset`. It is copied to a folder in my workspace, so read from there. On DGX06 I mounted the /raid/<criteo folder> to a folder inside the docker container.**

3. I use the same batch-size and most of the same functions, but I use a different network architecture as accelerating-tensorflow.ipynb . I think my network architecture is smaller, maybe thats the reason we see different behavior?

**could be. It might be related to DLRM architecture. But we think there is inefficient memory usage by the data loader at validation step or a bug**

**Comment by bschifferer** [comment]
> 1. Is the data to a local file system? If the data is read from the a "dataset", then this can be slow.
> 
> **No, on NGC, it is not read from the `dataset`. It is copied to a folder in my workspace, so read from there. On DGX06 I mounted the /raid/ to a folder inside the docker container.**

```--datasetid xxx:/dataset/ --workspace xxxxx:/workspace:RW```

Do you mean, the train and validation dataset is in /workspace? Can you copy the data to a local directory on the NGC instance (e.g. cp -R /workspace/data /tmp/data), first ? If the data is on a NGC dataset or NGC workspace, the instance will read the data from a network storage. I think it is faster to copy the data, first, to be sure, that the data is not read over network

**Comment by rnyak** [comment]
`Do you mean, the train and validation dataset is in /workspace? Can you copy the data to a local directory on the NGC instance (e.g. cp -R /workspace/data /tmp/data), first ?`

This is what I have done already.

**Comment by rnyak** [comment]
Do you mean, the train and validation dataset is in /workspace? Can you copy the data to a local directory on the NGC instance (e.g. cp -R /workspace/data /tmp/data), first ?

This is what I have done already.

**Comment by rnyak** [comment]
> > 1. Is the data to a local file system? If the data is read from the a "dataset", then this can be slow.
> > 
> > **No, on NGC, it is not read from the `dataset`. It is copied to a folder in my workspace, so read from there. On DGX06 I mounted the /raid/ to a folder inside the docker container.**
> 
> `--datasetid xxx:/dataset/ --workspace xxxxx:/workspace:RW`
> 
> Do you mean, the train and validation dataset is in /workspace? Can you copy the data to a local directory on the NGC instance (e.g. cp -R /workspace/data /tmp/data), first ? If the data is on a NGC dataset or NGC workspace, the instance will read the data from a network storage. I think it is faster to copy the data, first, to be sure, that the data is not read over network

@bschifferer 
This is what I have done already, i.e. `cp -R /workspace/train`. It is not read over network. Also note that, I have done the preprocessing offline with NVT, so, my data is also stored in a folder in my workspace. So it cannot read network storage I believe. As I said this behavior I shared above is same on a DGX06 system.

**Comment by jperez999** [comment]
@rnyak there is a lot to unpack here. Appreciate the abundance of information. I do need to ask for more specifics on your repro on DGX06 is this the ASE lab dgx06 machine or a separate machine that you are identifying as a dgx06? When I refer to the dgx06 system I am referring to a dgx-1 server with 8 - V100 32gb GPUs, I ask because of this statement "DGX06 system with single V100-32 GB GPU". Are we talking about the same resource? 

So this is a good place to start on spilling:
https://distributed.dask.org/en/latest/worker.html#memory-management

Also let me add a disclaimer, I am not sure this is the problem. This is just my best guess, I would have to be able to repro the problem to get a more concrete diagnosis.

**Comment by rnyak** [comment]
@jperez999 yes it is the same system: `dgx-1 server with 8 - V100 32gb GPUs,` but I am only using one single GPU there. I am using `joc:latest` container on `dgx-1 server`.

Julio, my main issue is that I cannot get the model validation acc metrics (auc and loss values) printed out even though I am using `validation_callback = KerasSequenceValidater(dataset)`. 

Strange behavior:
1. I can get them printed out with a small dataset
2. I used to get my validation set acc printed out even with larger dataset before NVT 0.3 release.

I believe there is a bug or something.

@gabrielspmoreira for viz.

**Comment by benfred** [comment]
This seems to be related to insufficient host / gpu memory - closing
